services/stats/stats.go

```diff
@@ -93,26 +97,59 @@ type RudderStatsT struct {
 	dontProcess bool
 }
 
+func getNewStatsdClientWithExpoBackoff(opts ...statsd.Option) {
+	maxWait := time.Minute * 10
+	bo := backoff.NewExponentialBackOff()
+	bo.MaxInterval = time.Minute
+	bo.MaxElapsedTime = maxWait
+	var err error
+	if err = backoff.Retry(func() error {
+		//statsd.New returns client even when err!=nil, although that client won't work.
+		//But, since we don't want getting new client to be a blocking call. So, setting it even with the non-working client.
+		//And, will be updated once we have `err==nil`, until then stats sent to telegraf will be dropped.
+		client, err = statsd.New(opts...)
+		if err != nil {
+			isMockClient = true
+			pkgLogger.Errorf("error while setting statsd client: %v", err)
+		}
+		return err
+	}, bo); err != nil {
+		if bo.NextBackOff() == backoff.Stop {
+			//if setup didn't succeed in bo.MaxElapsedTime, then we panic. Since, if `RSERVER_ENABLE_STATS` is `true`
+			// & we failed to complete statsd setup in maxElapsedTime then something is wrong. And, we don't want to loose metrics by ignoring the error.
+			panic(err)
+		}
+	}
+}
+
 //Setup creates a new statsd client
 func Setup() {
 	DefaultStats = &HandleT{}
 
 	if !statsEnabled {
 		return
 	}
-
-	var err error
 	conn = statsd.Address(statsdServerURL)
-	//TODO: Add tags by calling a function...
+	// since, we don't want setup to be a blocking call, creating a separate `go routine`` for retry to get statsd client.
+	var err error
+	//NOTE: this is to get atleast a dummy client, even in case of failure. So, that nil pointer error is not received when client is called.
 	client, err = statsd.New(conn, statsd.TagsFormat(getTagsFormat()), defaultTags())
 	if err != nil {
```

Comment:
I think we've got a logical error here. If the new statsd connection succeeds with the first attempt, then we're never reaching the call to `collectRuntimeStats(client)`.

Reply:
thanks for pointing this out, called `collectRuntimeStats` inside `if`  condition by mistake. Fixed.

Reply:
`statsd.New` returns a real client when `err==nil` & dummy client when `err!=nil`. That is the reason we are calling `statsd.New` once without retry mechanism. Since, retry mechanism is non-blocking. So, to avoid calling any of the `RudderStats` methods before getting a client (real/dummy), we are making a single blocking call request. So, that we don't get `nil` pointer error.